### PR TITLE
I was shamed into turning my email into a blog post:)

### PR DESCRIPTION
**Pull request by andreyk0**

### CODE REVIEW

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+author: Andrey Kartashov
+category: docker
+tags:
+- docker
+- osx
+- nfs
+- firewall
+---
+
+Mobile services team at Gilt uses Docker to both, build and run software.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: "The Mobile Services" and nix the comma.

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+In addition to the usual Docker benefits for software deployments moving toolchains to Docker has a few advantages:
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: deployments, moving

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+- it's easy to (re)create a development environment
+- environment is preserved in a stable binary form (libs, configs, CLI tools, etc, etc don't bit rot as main OS packages or OS itself evolve)
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: the environment

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+- easy to support multiple divergent environments where different versions of tools/libs are the default, e.g. java7/8, python, ruby, scala, etc
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: default;

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+We mostly develop on OSX, Docker is a Linux-specific tool, so we use docker-machine and VirtualBox to actually run it.
+To run toolchains we rely on having access to the host OS's filesystem, by default /Users is exposed in the Docker VM.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: wrap /Users in backticks so it shows up as code `/Users`

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+Unfortunately, default setup uses VBOXFS which is very slow. This can be really painful when building larger projects or relying on build steps that require a lot of IO, like [sbt-assembly](https://github.com/sbt/sbt-assembly) plugin.
+
+Here's a great [comparison of IO performance](http://mitchellh.com/comparing-filesystem-performance-in-virtual-machines).
+
+There's really no good solution for this problem at the moment but some folks have come up with a reasonable hack: use NFS.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: moment, but

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+One of them was even nice enough to wrap it up into a [shell script](https://github.com/adlogix/docker-machine-nfs) that "just works".
+
+
+Indeed, with NFS enabled project build times begin to approach "native" speeds, so it's tempting.
+One issue with it though is that NFS is a very old system, designed to work in trusted network environments, access is given to hosts, not authenticated users.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: The issue with NFS continues to be it's aging design and intent to function in trusted network environment where access is given to hosts, not to authenticated users.

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+While this is a reasonable access model for secure production networks it's hard to guarantee anything about random networks you may have to connect to with your laptop, having /Users exposed via NFS on un-trusted networks is a scary prospect.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: networks, it's

laptop, and having `/Users`

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+
+OSX has not one but two built-in firewalls.
+There's a simplified app-centric firewall available from Preferences panel.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: firewalls, and there's a simplified

firewall available in the System Preferences

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+Unfortunately all it can do is either block all nfs traffic (docker VM can't access your exported file system) or open up nfs traffic on all interfaces (insecure), so, it doesn't really work for this case.
+
+Fortunately, under the hood there's also a much more flexible built-in packet level firewall that can be configured.
+It's called PF (packet filter) and its main CLI tool is pfctl.
+Here's a nice [intro](https://pleiades.ucsc.edu/hyades/PF_on_Mac_OS_X).
+
+With that, one possible solution is to disable firewall in the Preferences panel and add this section at the end of the /etc/pf.conf file instead:
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: What what? :)

wrap /etc/pf.conf with inline backticks

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+		<string>-f</string>
+		<string>/etc/pf.conf</string>
+	</array>
+	<key>RunAtLoad</key>
+	<true/>
+</dict>
+</plist>
+```
+
+and configuring it to start by running
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: And

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+```sh
+sudo launchctl load -w /Library/LaunchDaemons/com.yourcompany.pfctl.plist
+```
+
+The main difference from /System/Library/LaunchDaemons/com.apple.pfctl.plist here is the addition of -E parameter.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: wrap the path in inline backticks

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+You can check that it starts by default after a reboot with
+
+```sh
+sudo pfctl -s info
+```
+
+and check the rules with
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: And

File: _posts/2016-03-10-osx-docker-firewall.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+It should be 'Enabled' and you should see the configured rules.
+
+You can verify your overall setup by running nmap from a different node against your laptop, e.g.
+
+```sh
+sudo nmap -P0 -sU  YOUR_IP
+sudo nmap -P0 -sT  YOUR_IP
+```
+
+and check for open ports.
```
Comment (shellscape) [inline _posts/2016-03-10-osx-docker-firewall.md]: And

**Comment by shellscape** [comment]
Great first draft. Post could really use a concluding paragraph. I'd also suggest considering which paragraphs you can combine, as just one sentence comprise a paragraph in a few places.

**Comment by krschultz** [comment]
LGTM

**Pull request merged**